src/client/debugger/jupyter/kernelDebugAdapter.ts

```diff
@@ -0,0 +1,331 @@
+                        if (source && source.path) {
+                            const cell = this.fileToCell.get(source.path);
+                            if (cell) {
+                                source.name = path.basename(cell.document.uri.path);
+                                if (cell.index >= 0) {
+                                    source.name += `, Cell ${cell.index + 1}`;
+                                }
+                                source.path = cell.document.uri.toString();
+                            }
+                        }
+                    });
+
+                    this.sendMessage.fire(msg.content);
+                };
+                control.onStdin = (msg) => {
+                    this.getMessageSourceAndHookIt(msg.content as DebugProtocol.ProtocolMessage, (source) => {
+                        if (source && source.path) {
+                            const cell = this.fileToCell.get(source.path);
+                            if (cell) {
+                                source.name = path.basename(cell.document.uri.path);
+                                if (cell.index >= 0) {
+                                    source.name += `, Cell ${cell.index + 1}`;
+                                }
+                                source.path = cell.document.uri.toString();
+                            }
+                        }
+                    });
+
+                    this.sendMessage.fire(msg.content);
+                };
+                this.messageListener.set(message.seq, control);
+            }
+        } else if (message.type === 'response') {
+            // responses of reverse requests
+            const response = debugResponse(message as DebugProtocol.Response);
+            this.jupyterSession.requestDebug({
+                seq: response.content.seq,
+                type: 'request',
+                command: response.content.command
+            });
+        } else {
+            // cannot send via iopub, no way to handle events even if they existed
+            console.assert(false, `Unknown message type to send ${message.type}`);
+        }
+    }
+
+    dispose() {
+        this.messageListener.forEach((ml) => ml.dispose());
+        this.messageListener.clear();
+    }
+
+    private async dumpCellsThatRanBeforeDebuggingBegan() {
+        this.cellMap.getCellsAnClearQueue(this.notebookDocument).forEach(async (cell) => {
+            await this.dumpCell(cell.document.uri.toString());
+        });
+    }
+
+    /**
+     * Dump content of given cell into a tmp file and return path to file.
+     */
+    private async dumpCell(uri: string): Promise<void> {
+        const cell = this.notebookDocument.getCells().find((c) => c.document.uri.toString() === uri);
+        if (cell) {
+            try {
+                const response = await this.session.customRequest('dumpCell', { code: cell.document.getText() });
```

Comment:
I guess this means the kernel has to support this request. I guess it's okay cause ipykernel is probably the only reasonable case for debugging, but maybe we should flag this somehow so we don't forget it when we try getting debugging working elsewhere.

Reply:
What do you mean by flagging? With a comment?

Reply:
Yeah with a comment here and maybe an issue? ipykernel supports 'dumpcell' but does xeus?